rfc/system/0000-flyteadmin-rbac.md

```diff
@@ -0,0 +1,203 @@
+# [RFC] FlyteAdmin RBAC + Project/Domain Isolation
+
+**Authors:**
+
+- @sovietaced
+
+## 1 Executive Summary
+
+We propose support for role based access control in Flyte Admin with support for project and domain level isolation/filtering.
+
+## 2 Motivation
+
+Support for authorization in Flyte Admin is minimal and may be unsuitable for production deployments. Flyte Admin only 
+supports blanket authorization which does not align with information security best practices like the  
+[principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). Flyte's project and domain 
+concepts seem like strong constructs for supporting multi-tenancy but there are no mechanisms in place to prevent users 
+of one tenant from accidentally or maliciously manipulating another tenant's executions, tasks, launch plans, etc.
+
+At Stack AV we have solved this problem in our fork and are looking to contribute our work back after several requests 
+from the community. 
+
+## 3 Proposed Implementation
+
+### High Level Design 
+
+![High Level Design](/rfc/images/rbac-high-level-black.png)
+
+This proposal introduces an authorization interceptor that will be executed after the authentication interceptor. The 
+authorization interceptor will consult the auth or authorization config to understand how it should resolve roles from 
+the identity context along with what authorization policies are tied to the user's roles. If the user has an authorization 
+policy that permits the incoming request the project and domain level scope will be extracted from the authorization 
+policy and injected into the authorization context for use in the repository layer. The request will hit the service layer which 
+will make requests to the repository layer. The repository layer will leverage some utility functions to 1) authorize 
+whether resources can be created in a project/domain and 2) generate some SQL filters to be injected into queries that 
+read/update/delete resources. 
+
+Ultimately the authorization interceptor will provide RPC level authorization and set context for later use. The changes to the repository layer 
+will provide resource level isolation within the target RPC. Below is a breakdown of the three primary components of the design. 
+
+#### Authorization Config 
+The authorization config will be used to configure the authorization behavior. The first thing it describe is a way to
+resolve roles from the user's identity context. Ideally this should be flexible enough to resolve a role from different
```

Comment:
for a [multi-cluster set-up](https://docs.flyte.org/en/latest/deployment/deployment/multicluster.html#scaling-beyond-kubernetes), where the control plane (admin) and data plane (propeller) reside on separate clusters, propeller authenticates with a set of client credentials against flyteadmin to send execution events and create executions (child workflows, e.g. calling a launch plan in code)

if I'm understanding correctly, in this case we could use the userID of the propeller credentials to permit deployment-wide RW access? this isn't quite userID, but I think your suggestion can very well support applicationID using this approach. My understanding is that it's a bit more tricky to add scopes & claims to applications rather than user identities

Reply:
We use a multi cluster setup and admittedly our implementation of this proposal actually has a separate config field for allow listing certain user IDs, which in our case are Okta client Ids for the propeller Okta app. This allow list gives them full access to all APIs and all projects/domains. This was more of an incremental thing I added when I realized very quickly that authorization broke propeller :)

When I was writing the proposal I realized that I could probably do away with user ID allow lists in our implementation and support role resolution more generally by allowing users to pick out which values from the token they want to potentially match with policies. So yeah, I think for Flyte workloads you'd just turn on the user ID or application ID resolution strategy and create a role like so, not having to deal with scopes at all.  

```
 - name: 0oahjhk34aUxGnWcZ0h7 # the names can even include things like okta app IDs
      rules:
        - name: "flyte propeller"
          methodPattern: ".*"
```   

I guess to be clear, the role resolution strategies configured will ultimately generate a set of strings and then try and match those strings against the policy map keys. 

But I'm open to doing something special for the Flyte workloads

Reply:
this proposal sounds like it unblocks propeller access! I'm curious how this works for supporting dynamically provisioned app credentials (besides just propeller, but for things like CICD, etc) since I believe Okta doesn't allow you to attach claims to applications unless you use custom claims in the authorization server which is a bit more involved